### ISSUE TITLE
What to prepared for a very basic Google Analytics integration into Baker 4.1?

### ISSUE THREAD

**Issue by ask91** [Simbul/baker#951]
Refers to issue #332 .
I'm trying to redocumentating the former instruction to add Google Analytics from the issue above. 
Discussed explicitly by @yujulio  @anjimi  @Simbul  and others:::

---

Added the GA files and the frameworks to Xcode project as per the GA basic overview.

imported GAI.h in AppDelegate.m and added the GA initialization code:

```
// Optional: automatically send uncaught exceptions to Google Analytics.
[GAI sharedInstance].trackUncaughtExceptions = YES;
// Optional: set Google Analytics dispatch interval to e.g. 20 seconds.
[GAI sharedInstance].dispatchInterval = 20;
// Optional: set debug to YES for extra debugging information.
[GAI sharedInstance].debug = YES;
// Create tracker instance.
id<GAITracker> tracker = [[GAI sharedInstance] trackerWithTrackingId:@"UA-TRACKING-ID"];
to - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
```

Replaced this line:

```
id<GAITracker> tracker = [[GAI sharedInstance] trackerWithTrackingId:@"UA-TRACKING-ID"];
with these two lines from the Advanced Config 
id newTracker = [[GAI sharedInstance]trackerWithTrackingId:@"UA-TRACKING-ID");
[GAI sharedInstance].defaultTracker = newTracker;
```

(with UA-TRACKING-ID changed to your GA app ID)

---

_yujulio commented 2 months ago:_
Try it looking at GA if there's any active app.

_anjimi commented 2 months ago:_
I discovered that I need to #import "GAI.h" in BakerBook.m and BakerViewController.m not just in AppDelegate.m; this stopped the errors I was getting: Use of undeclared identifier 'GAI'.

---

Here's what I have working now:

```
GA initialization code added to - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions in AppDelegate.m:

    // Optional: automatically send uncaught exceptions to Google Analytics.
    [GAI sharedInstance].trackUncaughtExceptions = YES;
    // Optional: set Google Analytics dispatch interval to e.g. 20 seconds.
    [GAI sharedInstance].dispatchInterval = 20;
    // Optional: set debug to YES for extra debugging information.
    [GAI sharedInstance].debug = YES;
    // Create tracker instance.
    id<GAITracker> tracker = [[GAI sharedInstance] trackerWithTrackingId:@"UA-TRACKING-ID"];
    // Set the new tracker as the default tracker, globally.
    [GAI sharedInstance].defaultTracker = tracker;
```

I'm now tracking book and page views as Screen Views rather than Events as this is more appropriate for page views; you can also track Screen Views in Real-Time in GA and don't have to wait for data on Events, which is nice :)

Here's my code in BakerBook.m:

```
- (id)initWithBookData:(NSDictionary *)bookData
{
    self = [super init];
    if (self && [self loadBookData:bookData]) {
        NSString *baseID = [self.title stringByAppendingFormat:@" %@", [self.url stringSHAEncoded]];
        self.ID = [self sanitizeForPath:baseID];

        NSLog(@"[BakerBook] 'book.json' parsed successfully. Book '%@' created with id '%@'.", self.title, self.ID);
        [[GAI sharedInstance].defaultTracker sendView:(self.title)];
        return self;
    }

    return nil;
}
```

This tracks book views, by book title, under Screen Views in GA.

And in BakerViewController.m:

```
- (BOOL)loadWebView:(UIWebView*)webView withPage:(int)page {

    NSString *path = [NSString stringWithString:[pages objectAtIndex:page - 1]];

    if ([[NSFileManager defaultManager] fileExistsAtPath:path]) {
        NSLog(@"[BakerView] Loading: %@", [[NSFileManager defaultManager] displayNameAtPath:path]);
        [webView loadRequest:[NSURLRequest requestWithURL:[NSURL fileURLWithPath:path]]];

        NSString *GAIpage = [book.title stringByAppendingPathComponent:[[path lastPathComponent] stringByDeletingPathExtension]];
        [[GAI sharedInstance].defaultTracker sendView:GAIpage];

        return YES;
    }
    return NO;
}
```

This tracks page views, by book title/page file name, under Screen Views in GA.

I added a couple lines to `- (void)loadModalWebView:(NSURL *)url` also in BakerViewController.m:

```
    NSString *GAIlink = [book.title stringByAppendingPathComponent:url.absoluteString];
    [[GAI sharedInstance].defaultTracker sendView:GAIlink];
```

This tracks modal views, by book title/modal link url, under Screen Views in GA.

---

And it's the basic question from me.

1) I downloaded the GA SDK for iOS [GoogleAnalyticsiOS_2-2.0beta4 as a folder]
but what to do then?

2) Where can i find GA framework to be included in my BakerBook Project just like @anjimi said at first line 
"Added the GA files and the frameworks to Xcode project as per the GA basic overview." ?

I would love to hear a complete documentation of setting the very basic GoogleAnalytics integration into Baker , since following instruction within a discussion is a bit confusing for me

Thanks to those who've had contribute @ issue #332 :)

**Comment by anjimi** [comment]
You can find the instructions for including the GA SDK here:
https://developers.google.com/analytics/devguides/collection/ios/v2/

**Comment by ghost** [comment]
Hi guys,

On this topic, are you planning implement GA in the master for 4.2 and have each user define the Google Analytics tracking code for their app in constants?

**Comment by folletto** [comment]
@mit1818 feature #332 is scheduled for Baker 4.2. 

We are finding however a few problems, because if we implement it, it needs to be abstract - so you could use either no analytics, one random analytics package, or Google Analytics.

**Comment by ghost** [comment]
Awesome. 
I hope you find a solution to your problem.

**Comment by ask91** [comment]
i followed what you've tried above just for testing and see how analytic works..
Anyway, i put the all GoogleAnalytic files::
GAI.h
GAITracker.h
GAITrackedViewController.h
GAITransaction.h
GAITransactionItem.h
libGoogleAnalytics.a

 into my project by dragging them to Project part in XCODE 
and the file hierarchy shows that those files equal to folders books BakerShelf BakerView Baker 

does it right or wrong?

**Comment by anjimi** [comment]
I'm not quite sure what you mean, but yes you add the files by dragging them to your Xcode project. 
I dragged the folder downloaded from google called 'Google Analytics' which contains those files, so in my project the GA files are all in a subfolder called 'Google Analytics'.

**Comment by folletto** [comment]
Since it's now implemented in the current master and future 4.2, I'm closing this old issue. :)
